twoliter/src/lock.rs

```diff
@@ -464,38 +590,75 @@ impl Lock {
         })
     }
 
+    #[instrument(level = "trace", skip(image), fields(image = %image))]
     async fn find_kit(
         image_tool: &ImageTool,
         vendor: &Vendor,
         image: &LockedImage,
     ) -> Result<ImageMetadata> {
+        debug!(kit_image = %image, "Searching for kit");
         let manifest_list: ManifestListView = serde_json::from_slice(image.manifest.as_slice())
             .context("failed to deserialize manifest list")?;
-        let mut encoded_metadata: Option<String> = None;
-        for manifest in manifest_list.manifests.iter() {
-            let image_uri = format!("{}/{}@{}", vendor.registry, image.name, manifest.digest);
-
-            // Now we want to fetch the metadata from the OCI image config
-            let config = image_tool.get_config(image_uri.as_str()).await?;
-            let encoded = config
-                .labels
-                .get("dev.bottlerocket.kit.v1")
-                .context("no metadata stored on image, this image appears to not be a kit")?;
-            if let Some(metadata) = encoded_metadata.as_ref() {
-                ensure!(
-                    encoded == metadata,
-                    "metadata does match between images in manifest list"
+        trace!(manifest_list = ?manifest_list, "Deserialized manifest list");
+
+        debug!("Extracting kit metadata from OCI image");
+        let embedded_kit_metadata = stream::iter(manifest_list.manifests)
+            .map(|manifest| async move {
+                let image_uri = format!("{}/{}@{}", vendor.registry, image.name, manifest.digest);
+                EncodedKitMetadata::try_from_image(&image_uri, image_tool).await
+            })
```

Comment:
Is the Docker image tool going to be robust for parallel invocations?

Reply:
It seems to support it (based on my own testing and a lack of warnings found online), but since a pull can result in many parallel layer pulls, 8 is probably too generous of a limit.

I also don't know if the parallel pulls collapse common layers into a single request (I'm guessing not), but the images won't really have layers in common anyways, since they are different architectures.

My real goal here was to separate the kit metadata fetch from the check that the metadata was consistent across images as a clarity change. The parallelism was added as a bonus. I wouldn't be opposed to removing it. For now, I'll just lower it.